lib/Feature/FeatureModelWriter.cpp

```diff
@@ -0,0 +1,204 @@
+#include "vara/Feature/FeatureModelWriter.h"
+#include "vara/Feature/FeatureModel.h"
+
+#include <libxml/xmlwriter.h>
+
+#define CHECK_RC if (rc < 0) { return rc; }
+
+namespace vara::feature {
+
+
+int FeatureModelXmlWriter::writeFeatureModel(string path) {
+    int rc;
+    xmlTextWriterPtr writer;
+    xmlChar *tmp;
+
+    writer = xmlNewTextWriterFilename(path.data(), 0);
+    if (writer == nullptr) {
+        return false;
+    }
+
+    rc = writeFeatureModel(writer);
+    xmlFreeTextWriter(writer);
+    return rc;
+}
+
+
+int FeatureModelXmlWriter::writeFeatureModel(char **doc) {
+    int rc;
+    xmlTextWriterPtr writer;
+    xmlBufferPtr buf;
+
+    buf = xmlBufferCreate();
+    writer = xmlNewTextWriterMemory(buf, 0);
+    rc = writeFeatureModel(writer);
+
+    xmlFreeTextWriter(writer);
+    // copy document into char buffer
+    *doc = static_cast<char *>(calloc(xmlBufferLength(buf), 1));
+    memcpy(*doc, xmlBufferContent(buf), xmlBufferLength(buf));
+
+    xmlBufferFree(buf);
+    return rc;
+}
+
+int FeatureModelXmlWriter::writeFeatureModel(xmlTextWriterPtr writer) {
+    int rc;
+    rc = xmlTextWriterStartDocument(writer, "1.0", "UTF-8", nullptr);
+    CHECK_RC
+    rc = writeVm(writer);
+    // report failure/success
+    return rc;
+}
+
+int FeatureModelXmlWriter::writeVm(xmlTextWriterPtr writer) {
+    int rc;
+    // open vm node
+    rc = xmlTextWriterStartElement(writer, VM);
+    CHECK_RC
+    rc = xmlTextWriterWriteAttribute(writer, NAME, BAD_CAST fm.getName().data());
+    CHECK_RC
+    rc = xmlTextWriterWriteAttribute(writer, ROOT, BAD_CAST fm.getPath().string().data());
+    CHECK_RC
+
+    // write BinaryFeatures
+    rc = writeBinaryFeatures(writer);
+    CHECK_RC
+
+    // write NumericFeatures
+    rc = writeNumericFeatures(writer);
+    CHECK_RC
+
+    // write Constraints
+    rc = writeBooleanConstraints(writer);
+    CHECK_RC
+
+    rc = xmlTextWriterEndElement(writer); // VM
+    return rc;
+}
+
+int FeatureModelXmlWriter::writeBinaryFeatures(xmlTextWriterPtr writer) {
+    int rc;
+    rc = xmlTextWriterStartElement(writer, BINARYOPTIONS);
+    CHECK_RC
+    for (Feature *f : fm.features()) {
+        if (f->getKind() == Feature::FeatureKind::FK_BINARY) {
+            rc = writeFeature(writer, *f);
+        }
+    }
+
+    rc = xmlTextWriterEndElement(writer); // BINARYOPTIONS
+    return rc;
+}
+
+int FeatureModelXmlWriter::writeNumericFeatures(xmlTextWriterPtr writer) {
+    int rc;
+    rc = xmlTextWriterStartElement(writer, NUMERICOPTIONS);
+    CHECK_RC
+    for (Feature *f : fm.features()) {
+      if (f->getKind() == Feature::FeatureKind::FK_NUMERIC) {
+        rc = writeFeature(writer, *f);
+      }
+    }
+
+    rc = xmlTextWriterEndElement(writer); // NUMERICOPTIONS
+    return rc;
+}
+
+int FeatureModelXmlWriter::writeBooleanConstraints(xmlTextWriterPtr writer) {
```

Comment:
```suggestion
Error FeatureModelXmlWriter::writeBooleanConstraints(xmlTextWriterPtr writer) {
```
As discussed, we should change this to some error wrapper